src/io/simulator/simulator_handle.hpp

```diff
@@ -0,0 +1,453 @@
+        }
+      }
+    }
+
+    if (blocked_servers < server_addresses_.size()) {
+      // we only need to advance the simulator when all
+      // servers have reached a quiescent state, blocked
+      // on their own futures or receive methods.
+      return false;
+    }
+
+    stats_.simulator_ticks++;
+
+    cv_.notify_all();
+
+    TimeoutPromisesPastDeadline();
+
+    if (in_flight_.empty()) {
+      // return early here because there are no messages to schedule
+
+      // We tick the clock forward when all servers are blocked but
+      // there are no in-flight messages to schedule delivery of.
+      std::poisson_distribution<> time_distrib(50);
+      uint64_t clock_advance = time_distrib(rng_);
+      cluster_wide_time_microseconds_ += clock_advance;
+
+      MG_ASSERT(cluster_wide_time_microseconds_ < config_.abort_time,
+                "Cluster has executed beyond its configured abort_time, and something may be failing to make progress "
+                "in an expected amount of time.");
+      return true;
+    }
+
+    if (config_.scramble_messages) {
+      // scramble messages
+      std::uniform_int_distribution<size_t> swap_distrib(0, in_flight_.size() - 1);
+      size_t swap_index = swap_distrib(rng_);
+      std::swap(in_flight_[swap_index], in_flight_.back());
+    }
+
+    auto [to_address, opaque_message] = std::move(in_flight_.back());
+    in_flight_.pop_back();
+
+    std::uniform_int_distribution<int> drop_distrib(0, 99);
+    int drop_threshold = drop_distrib(rng_);
+    bool should_drop = drop_threshold < config_.drop_percent;
+
+    if (should_drop) {
+      stats_.dropped_messages++;
+    }
+
+    PromiseKey promise_key{.requester_address = to_address,
+                           .request_id = opaque_message.request_id,
+                           .replier_address = opaque_message.from_address};
+
+    if (promises_.contains(promise_key)) {
+      // complete waiting promise if it's there
+      DeadlineAndOpaquePromise dop = std::move(promises_.at(promise_key));
+      promises_.erase(promise_key);
+
+      bool normal_timeout = config_.perform_timeouts && (dop.deadline < cluster_wide_time_microseconds_);
+
+      if (should_drop || normal_timeout) {
+        stats_.timed_out_requests++;
+        dop.promise.TimeOut();
+      } else {
+        stats_.total_responses++;
+        dop.promise.Fill(std::move(opaque_message));
+      }
+    } else if (should_drop) {
+      // don't add it anywhere, let it drop
+    } else {
+      // add to can_receive_ if not
+      const auto &[om_vec, inserted] = can_receive_.try_emplace(to_address, std::vector<OpaqueMessage>());
+      om_vec->second.emplace_back(std::move(opaque_message));
+    }
+
+    return true;
+  }
+
+  void ShutDown() {
+    std::unique_lock<std::mutex> lock(mu_);
+    should_shut_down_ = true;
+    cv_.notify_all();
+  }
+
+  bool ShouldShutDown() {
+    std::unique_lock<std::mutex> lock(mu_);
+    return should_shut_down_;
+  }
```

Comment:
You can mark `mu_` mutable, so this function can be turned into a `const` member function.